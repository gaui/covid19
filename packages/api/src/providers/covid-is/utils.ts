/* eslint-disable @typescript-eslint/ban-ts-ignore */
import cheerio from 'cheerio';
import * as R from 'ramda';
import {
  InfographicDataWindow,
  InfographicData,
  ElementsContentContent,
  Covid19Stats
} from './types';

const labelMap = {
  'staðfest smit': 'cases',
  'í sóttkví': 'quarantineIn',
  'lokið sóttkví': 'quarantinePost',
  'Í einangrun': 'isolated',
  'lokið einangrun': 'isolatedPost',
  'á sjúkrahúsi': 'hospitalized',
  'á gjörgæslu': 'critical',
  'náð bata': 'recovered',
  'sýni, innanlands': 'samples',
  'sýni, landamæri': 'samplesBorder',
};

export const labelMapper: (array: [string, number][]) => Covid19Stats[][] = (
  array: [string, number][]
) => array.map(([label, count]) => [labelMap[label], count]);

const getValueFromHTML: (val: string) => string = R.tryCatch<string>(
  v => cheerio(cheerio.load(v).html()).text(),
  () => {
    return '';
  }
);

const getPath: (x: InfographicData) => unknown[] = R.compose(
  R.map(R.path(['data', 0, 0])),
  R.filter(R.propEq('chart_type_nr', 23)),
  R.map(R.path(['props', 'chartData'])),
  R.filter(R.propEq('type', 'CHART')),
  Object.values,
  R.pathOr({} as ElementsContentContent, [
    'elements',
    'content',
    'content',
    'entities'
  ])
);

const isNotEmpty = R.complement(R.isEmpty);
const filterValues = R.filter(R.allPass([isNotEmpty]));

const removeSymbols = (symbols: string[]) =>
  R.replace(RegExp(`[${symbols.join('.')}]+`, 'g'), '');
const convertToNumber = Number;
const isNumber: (value: never) => boolean = R.compose(Number.isFinite, Number);
const convertValue = R.ifElse(
  isNumber,
  R.compose(convertToNumber, removeSymbols(['.'])),
  R.identity
);

export const filter: (data: InfographicData) => [string, number][] = R.compose(
  // @ts-ignore
  R.filter(R.any(isNumber)),
  R.map(R.map(convertValue)),
  R.map(filterValues)
);

const parseHTML: (rawData: string) => InfographicData = (rawData: string) => {
  const $ = cheerio.load(rawData);
  const html = $('script:not([src])')
    .eq(3)
    .html() as string;
  const window = {} as InfographicDataWindow;
  eval(html);
  return window.infographicData;
};

const parseObject = R.compose(
  R.map(R.map(getValueFromHTML)),
  // @ts-ignore
  R.map(R.reverse),
  R.map(R.slice(0, 2)),
  getPath
);

export const parse = R.compose(parseObject, parseHTML);
