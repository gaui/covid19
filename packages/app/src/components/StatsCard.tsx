/* eslint-disable @typescript-eslint/no-var-requires */
import React from 'react';
import { StatsCardProps } from '../types/components';

const StatsCard = ({ icon, count, ...props }: StatsCardProps) => {
  const Icon = () => icon;

  return (
    <div {...props}>
      <Icon />
      <div>{count == null ? '-' : count}</div>
    </div>
  );
};

StatsCard.displayName = 'StatsCard';

export default StatsCard;
