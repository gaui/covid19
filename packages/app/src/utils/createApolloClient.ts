import ApolloClient from 'apollo-boost';
import unfetch from 'unfetch';

let apolloClient: ApolloClient<unknown>;

export const createApolloClient = () => {
  if (!apolloClient)
    apolloClient = new ApolloClient({
      uri: process.env.COVID_API_URL,
      fetch: fetch || unfetch
    });

  return apolloClient;
};
