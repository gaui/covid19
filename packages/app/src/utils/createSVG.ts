import React from 'react';

export const createSVG = (svg: string) => {
  const svgContainer = document.createElement('div');
  svgContainer.innerHTML = svg;
  const svgElement = svgContainer.firstElementChild as Element;
  const props = Object.fromEntries(
    Array.from(svgElement.attributes).map(x => [x.name, x.value])
  );

  // eslint-disable-next-line react/display-name
  return () =>
    React.createElement('svg', {
      ...props,
      dangerouslySetInnerHTML: { __html: svgElement.innerHTML }
    });
};
