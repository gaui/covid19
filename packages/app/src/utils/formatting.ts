export const thousandSeperator = (num: string | number) =>
  new Intl.NumberFormat('is-IS').format(Number(num));
