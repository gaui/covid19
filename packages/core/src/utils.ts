import unfetch from 'isomorphic-unfetch';

const fetch = unfetch.bind(null);

export default fetch;
