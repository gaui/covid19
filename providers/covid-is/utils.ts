import R from 'ramda';

const labelMap = {
  'staðfest smit': 'infected',
  sýni: 'samples',
  'í sóttkví': 'quarantined',
  'í einangrun': 'isolated',
  'á sjúkrahúsi': 'hospitalized'
};

const isChartType = ([_, o]: [string, Entity]) => o.type === 'CHART';
const getObjectChildTree = ([_, o]: [string, Entity]) => [
  R.path(['props', 'chartData', 'data', 0, 0], o)
];
const isValidFormat = ([o]: [any]) =>
  o.length === 3 && typeof o[0] === 'string';
const createStructure = ([o]: [Entity]) => [
  labelMap[o[1]],
  Number(o[0].replace(/<[^>]*>?|\./gm, ''))
];

export { isChartType, getObjectChildTree, isValidFormat, createStructure };
