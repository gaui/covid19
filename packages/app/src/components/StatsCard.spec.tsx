import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import StatsCard from './StatsCard';
import Cases from '../svg/cases.svg';
import { createSVG } from '../utils/createSVG';

const CasesSVG = createSVG(Cases);

it('renders a StatsCard for infection with count of 5', async () => {
  const { getByTestId } = render(
    <StatsCard
      icon={<CasesSVG />}
      title="Total cases"
      count={5}
      data-testid="StatsCard5"
    />
  );

  const node = await getByTestId('StatsCard5');

  expect(node.lastChild).toMatchSnapshot();
});
