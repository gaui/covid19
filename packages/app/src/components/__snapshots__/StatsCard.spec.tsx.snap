// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a StatsCard for infection with count of 5 1`] = `
<div
  class="stats"
>
  5
</div>
`;
